Source/WebGPU/WGSL/Overload.cpp

```diff
@@ -135,27 +135,31 @@ Type* OverloadResolver::resolve()
     return selectedCandidate->result;
 }
 
-Type* OverloadResolver::materialize(const AbstractType& abstractType) const
+std::optional<Type*> OverloadResolver::materialize(const AbstractType& abstractType) const
 {
     return WTF::switchOn(abstractType,
-        [&](Type* type) -> Type* {
-            return type;
+        [&](Type* type) -> std::optional<Type*> {
+            return { type };
         },
-        [&](TypeVariable variable) -> Type* {
-            auto* resolvedType = resolve(variable);
-            ASSERT(resolvedType);
-            return resolvedType;
+        [&](TypeVariable variable) -> std::optional<Type*> {
+            if (auto* resolvedType = resolve(variable))
+                return { resolvedType };
+            return std::nullopt;
         },
-        [&](const AbstractVector& vector) -> Type* {
-            auto* element = materialize(vector.element);
-            auto size = materialize(vector.size);
-            return m_types.vectorType(element, size);
+        [&](const AbstractVector& vector) -> std::optional<Type*> {
+            if (auto element = materialize(vector.element)) {
+                auto size = materialize(vector.size);
+                return { m_types.vectorType(*element, size) };
+            }
+            return std::nullopt;
         },
-        [&](const AbstractMatrix& matrix) -> Type* {
-            auto* element = materialize(matrix.element);
-            auto columns = materialize(matrix.columns);
-            auto rows = materialize(matrix.rows);
-            return m_types.matrixType(element, columns, rows);
+        [&](const AbstractMatrix& matrix) -> std::optional<Type*> {
+            if (auto element = materialize(matrix.element)) {
```

Comment:
I don't understand how `materialize(matrix.element)` knows that we're in the middle of materializing a matrix, so only some types are acceptable. There doesn't seem to be any connection between the nested materialize calls to provide context between them. 🤔

Reply:
`AbstractMatrix` can only be constructed from `TypeDeclarations.rb`, and the declaration provides constraints on the variable (if you look at all the matrix types, they all take variables constrained to `Float`). We materialize variables by looking them up in the type substitution dictionary, but we only insert a substitution if it satisfies the constraints, so by the time we get to materialization this has already been handled.

There's another part to this, which is validating user-provided types, which is not implemented yet (e.g. we should, but still don't, reject `let x : mat2x2<i32>;`). This will happen in `TypeCheck.cpp` though, so it's not part of overload resolution.